File: src/main/java/hiiragi283/material/api/capability/fluid/HiiragiFluidTank.java
package hiiragi283.material.api.capability.fluid;

import hiiragi283.material.api.capability.IOControllable;
import hiiragi283.material.util.HiiragiUtil;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraftforge.common.util.INBTSerializable;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.fluids.FluidTank;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.annotation.ParametersAreNonnullByDefault;
import java.util.Optional;

@ParametersAreNonnullByDefault
public class HiiragiFluidTank extends FluidTank implements IOControllable, INBTSerializable<NBTTagCompound> {

    @NotNull
    protected final IOControllable.Type ioType;
    @Nullable
    protected final TileEntity tile;

    public HiiragiFluidTank(int capacity, IOControllable.Type ioType) {
        this(capacity, ioType, null);
    }

    public HiiragiFluidTank(int capacity, IOControllable.Type ioType, @Nullable TileEntity tile) {
        super(capacity);
        this.ioType = ioType;
        this.tile = tile;
    }

    public void clear() {
        this.fluid = null;
    }

    public Optional<FluidStack> getFluidOptional() {
        return this.fluid != null ? Optional.of(this.fluid) : Optional.empty();
    }

    public boolean isEmpty() {
        return this.fluid == null;
    }

    @Override
    protected void onContentsChanged() {
        if (tile != null) {
            tile.markDirty();
        }
    }

    //    IOControllable    //

    @NotNull
    @Override
    public Type getIOType() {
        return ioType;
    }

    //    INBTSerializable    //

    @Override
    public NBTTagCompound serializeNBT() {
        var tag = new NBTTagCompound();
        tag.setInteger(HiiragiUtil.CAPACITY, getCapacity());
        getFluidOptional().ifPresent(stack -> tag.setTag(HiiragiUtil.FLUID, stack.writeToNBT(new NBTTagCompound())));
        return tag;
    }

    @Override
    public void deserializeNBT(NBTTagCompound nbt) {
        this.setCapacity(nbt.getInteger(HiiragiUtil.CAPACITY));
        this.setFluid(FluidStack.loadFluidStackFromNBT(nbt.getCompoundTag(HiiragiUtil.FLUID)));
    }

}

File: src/main/kotlin/hiiragi283/material/api/capability/fluid/HiiragiFluidTank.kt
package hiiragi283.material.api.capability.fluid

import hiiragi283.material.api.capability.IOControllable
import hiiragi283.material.util.HiiragiNBTUtil
import hiiragi283.material.util.getStringOrNull
import net.minecraft.nbt.NBTTagCompound
import net.minecraftforge.common.util.INBTSerializable
import net.minecraftforge.fluids.FluidStack
import net.minecraftforge.fluids.FluidTank

open class HiiragiFluidTank(
    capacity: Int,
    override var ioType: IOControllable.Type = IOControllable.Type.GENERAL
) : FluidTank(capacity), IOControllable, INBTSerializable<NBTTagCompound> {

    override fun canFillFluidType(fluid: FluidStack?): Boolean =
        getFluid() == null || getFluid()!!.isFluidEqual(fluid)

    override fun canDrainFluidType(fluid: FluidStack?): Boolean =
        getFluid() !== null && getFluid()!!.isFluidEqual(fluid)

    //    INBTSerializable    //

    override fun serializeNBT(): NBTTagCompound {
        return NBTTagCompound().also { tag: NBTTagCompound ->
            tag.setInteger(HiiragiNBTUtil.CAPACITY, getCapacity())
            getFluid()?.writeToNBT(NBTTagCompound())?.let { tagFluid: NBTTagCompound ->
                tag.setTag(HiiragiNBTUtil.FLUID, tagFluid)
            }
            tag.setString(HiiragiNBTUtil.IO_TYPE, ioType.name)
        }
    }

    override fun deserializeNBT(nbt: NBTTagCompound) {
        setCapacity(nbt.getInteger(HiiragiNBTUtil.CAPACITY))
        setFluid(FluidStack.loadFluidStackFromNBT(nbt.getCompoundTag(HiiragiNBTUtil.FLUID)))
        nbt.getStringOrNull(HiiragiNBTUtil.IO_TYPE)?.let { name: String -> ioType = IOControllable.Type.valueOf(name) }
    }

}